crates/api_common/src/request.rs

```diff
@@ -53,43 +54,53 @@ pub async fn fetch_link_metadata(url: &Url, context: &LemmyContext) -> LemmyResu
   // We only fetch the first 64kB of data in order to not waste bandwidth especially for large
   // binary files
   let bytes_to_fetch = 64 * 1024;
-  let response = context
-    .client()
-    .get(url.as_str())
-    // we only need the first chunk of data. Note that we do not check for Accept-Range so the
-    // server may ignore this and still respond with the full response
-    .header(RANGE, format!("bytes=0-{}", bytes_to_fetch - 1)) /* -1 because inclusive */
-    .send()
-    .await?
-    .error_for_status()?;
+  let mut response = Some(
```

Comment:
That's not what optionals are for. That means you need to clone / copy the response into a new variable, or use references if possible.

Reply:
I fixed it by using `else if` to directly prevent ownership from being moved more than once in the same branch

```
let response = context
    .client()
    .get(url.as_str())
    // we only need the first chunk of data. Note that we do not check for Accept-Range so the
    // server may ignore this and still respond with the full response
    .header(RANGE, format!("bytes=0-{}", bytes_to_fetch - 1)) /* -1 because inclusive */
    .send()
    .await?
    .error_for_status()?;

  let mut content_type: Option<Mime> = response
    .headers()
    .get(CONTENT_TYPE)
    .and_then(|h| h.to_str().ok())
    .and_then(|h| h.parse().ok());

  let mut opengraph_data = Default::default();

  if let Some(c) = &content_type {
    if (c.type_() == mime::TEXT && c.subtype() == mime::HTML)
	    ||
	    // application/xhtml+xml is a subset of HTML
	    (c.type_() == mime::APPLICATION && c.subtype() == "xhtml")
    {
      // Can't use .text() here, because it only checks the content header, not the actual bytes
      // https://github.com/LemmyNet/lemmy/issues/1964
      // So we want to do deep inspection of the actually returned bytes but need to be careful not
      // spend too much time parsing binary data as HTML

      // only take first bytes regardless of how many bytes the server returns
      let html_bytes = collect_bytes_until_limit(response, bytes_to_fetch).await?;
      opengraph_data = extract_opengraph_data(&html_bytes, url)
        .map_err(|e| info!("{e}"))
        .unwrap_or_default();
    }
    // If a server is serving `application/octet-stream`, it's likely a mistake,
    // so we try to guess the file type from its magic number.
    else if c.subtype() == "octet-stream" {
      // Don't need to fetch as much data for this as we do with opengraph
      let octet_bytes = collect_bytes_until_limit(response, 512).await?;
      content_type = infer::get(&octet_bytes).map_or(content_type, |t| t.mime_type().parse().ok());
    }
  }
```

Reply:
I landed on this solution because `reqwest::Response` doesn't implement Clone/Copy and needs to consume itself to get the body, but dullbananas' solution is cleaner than mine.